infra/gravity/config.go

```diff
@@ -0,0 +1,123 @@
+package gravity
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/gravitational/robotest/infra"
+
+	"github.com/go-yaml/yaml"
+	"github.com/stretchr/testify/require"
+	"gopkg.in/go-playground/validator.v9"
+)
+
+// ProvisionerConfig defines parameters required to provision hosts
+// CloudProvider, AWS, Azure, ScriptPath and InstallerURL
+type ProvisionerConfig struct {
+	// DeployTo defines cloud to deploy to
+	CloudProvider string `yaml:"cloud_provider" validate:"required,eq=aws|eq=azure"`
+	// AWS defines AWS connection parameters
+	AWS *infra.AWSConfig `yaml:"aws"`
+	// Azure defines Azure connection parameters
+	Azure *infra.AzureConfig `yaml:"azure"`
+
+	// ScriptPath is the path to the terraform script or directory for provisioning
+	ScriptPath string `yaml:"script_path" validate:"required"`
+	// InstallerURL is AWS S3 URL with the installer
+	InstallerURL string `yaml:"installer_url" validate:"required,url`
+
+	// Tag will group provisioned resources under for easy removal afterwards
+	tag string `validate:"required"`
+	// StateDir defines base directory where to keep state (i.e. terraform configs/vars)
+	stateDir string `validate:"required"`
+	// NodeCount defines amount of nodes to be provisioned
+	nodeCount uint `validate:"gte=1"`
+	// OS defines one of supported operating systems
+	os string `validate:"required,eq=ubuntu|eq=debian|eq=rhel|eq=centos"`
+}
+
+// LoadConfig loads essential parameters from YAML
+func LoadConfig(t *testing.T, configFile, stateDir, tag string) *ProvisionerConfig {
+	require.NotEmpty(t, configFile, "config file")
+	f, err := os.Open(configFile)
+	require.NoError(t, err, configFile)
+	defer f.Close()
+
+	configBytes, err := ioutil.ReadAll(f)
+	require.NoError(t, err)
+
+	cfg := ProvisionerConfig{}
+	err = yaml.Unmarshal(configBytes, &cfg)
+	require.NoError(t, err, configFile)
+
+	if cfg.tag = tag; cfg.tag == "" {
+		now := time.Now()
+		cfg.tag = fmt.Sprintf("RT-%02d%02d-%02d%02d",
+			now.Month(), now.Day(), now.Hour(), now.Minute())
+	}
+
+	if cfg.stateDir = stateDir; cfg.stateDir == "" {
+		cfg.stateDir, err = ioutil.TempDir("", "robotest")
+		require.NoError(t, err, "tmp dir")
+	}
+
+	cfg.stateDir = filepath.Join(cfg.stateDir, cfg.tag)
+
+	return &cfg
```

Comment:
yep, I had some functions which returned it together with error, so although functions are gone, the pointer remains in case the need will arise again.